src/i2c/src/i2c.rs

```diff
@@ -23,10 +111,237 @@ pub struct I2cBackend {
     adapter_map: [u32; MAX_I2C_VDEV],
 }
 
+fn vi2c_set_client_addr(fd: i32, addr: usize) -> bool {
+    let ret = unsafe { ioctl(fd, I2C_SLAVE, addr as c_ulong) };
+    if ret == -1 {
+        println!("Failed to set client addr to {:x}", addr);
+        false
+    } else {
+        true
+    }
+}
+
+fn i2c_xfer(adapter: &I2cAdapter, reqs: &mut [I2cReq]) -> bool {
+    let mut msgs: Vec<I2cMsg> = Vec::with_capacity(reqs.len());
+    let len = reqs.len();
+    let addr = reqs[0].addr;
+
+    for req in reqs {
+        msgs.push(I2cMsg {
+            addr: req.addr,
+            flags: req.flags,
+            len: req.len,
+            buf: req.buf.as_mut_ptr(),
+        });
+    }
+
+    let data = I2cRdwrIoctlData {
+        msgs: msgs.as_mut_ptr(),
+        nmsgs: len as u32,
+    };
+
+    let ret = unsafe { ioctl(adapter.fd, I2C_RDWR, &data) };
+    if ret == -1 {
+        println!("Failed to transfer i2c data to client addr to {:x}", addr);
+        false
+    } else {
+        true
+    }
+}
+
+/// Based on Linux's drivers/i2c/i2c-core-smbus.c:i2c_smbus_xfer_emulated().
+///
+/// This function tries to reverse what Linux does, only support basic modes
+/// (up to word transfer).
+fn smbus_xfer(adapter: &I2cAdapter, reqs: &mut [I2cReq]) -> bool {
+    let mut data = I2cSmbusData {
+        block: [0; I2C_SMBUS_BLOCK_MAX + 2],
+    };
+    let read_write: u8;
+    let size: u32;
+
+    match reqs.len() {
+        1 => {
+            if (reqs[0].flags & I2C_M_RD) != 0 {
+                if reqs[0].len > 1 {
+                    println!(
+                        "Incorrect message length for read operation: {}",
+                        reqs[0].len
+                    );
+                    return false;
```

Comment:
This should be replaced with an `Error` instead.

Reply:
I am not sure if I was able to follow this well. Were you talking about the return statement or the println!() as well ? Can you give an example on how you want to see this ?